File: packages/expo-updates/ios/EXUpdates/SelectionPolicy/EXUpdatesSelectionPolicyFactory.m
//  Copyright © 2021 650 Industries. All rights reserved.

#import <EXUpdates/EXUpdatesLauncherSelectionPolicyFilterAware.h>
#import <EXUpdates/EXUpdatesLoaderSelectionPolicyFilterAware.h>
#import <EXUpdates/EXUpdatesReaperSelectionPolicyFilterAware.h>
#import <EXUpdates/EXUpdatesSelectionPolicyFactory.h>

NS_ASSUME_NONNULL_BEGIN

@implementation EXUpdatesSelectionPolicyFactory

+ (EXUpdatesSelectionPolicy *)filterAwarePolicyWithRuntimeVersion:(NSString *)runtimeVersion
{
  return [[EXUpdatesSelectionPolicy alloc]
          initWithLauncherSelectionPolicy:[[EXUpdatesLauncherSelectionPolicyFilterAware alloc] initWithRuntimeVersion:runtimeVersion]
          loaderSelectionPolicy:[[EXUpdatesLoaderSelectionPolicyFilterAware alloc] init]
          reaperSelectionPolicy:[[EXUpdatesReaperSelectionPolicyFilterAware alloc] init]];
}

+ (EXUpdatesSelectionPolicy *)filterAwarePolicyWithRuntimeVersions:(NSArray<NSString *> *)runtimeVersions
{
  return [[EXUpdatesSelectionPolicy alloc]
          initWithLauncherSelectionPolicy:[[EXUpdatesLauncherSelectionPolicyFilterAware alloc] initWithRuntimeVersions:runtimeVersions]
          loaderSelectionPolicy:[[EXUpdatesLoaderSelectionPolicyFilterAware alloc] init]
          reaperSelectionPolicy:[[EXUpdatesReaperSelectionPolicyFilterAware alloc] init]];
}

@end

NS_ASSUME_NONNULL_END

File: packages/expo-updates/ios/EXUpdates/SelectionPolicy/EXUpdatesSelectionPolicyFactory.swift
//  Copyright © 2019 650 Industries. All rights reserved.

import Foundation

/**
 * Factory class to ease the construction of [SelectionPolicy] objects whose three methods all use
 * the same ordering policy.
 */
@objcMembers
public final class EXUpdatesSelectionPolicyFactory: NSObject {
  public static func filterAwarePolicy(withRuntimeVersion runtimeVersion: String) -> EXUpdatesSelectionPolicy {
    return EXUpdatesSelectionPolicy.init(
      launcherSelectionPolicy: EXUpdatesLauncherSelectionPolicyFilterAware.init(runtimeVersion: runtimeVersion),
      loaderSelectionPolicy: EXUpdatesLoaderSelectionPolicyFilterAware(),
      reaperSelectionPolicy: EXUpdatesReaperSelectionPolicyFilterAware()
    )
  }

  public static func filterAwarePolicy(withRuntimeVersions runtimeVersions: [String]) -> EXUpdatesSelectionPolicy {
    return EXUpdatesSelectionPolicy.init(
      launcherSelectionPolicy: EXUpdatesLauncherSelectionPolicyFilterAware.init(runtimeVersions: runtimeVersions),
      loaderSelectionPolicy: EXUpdatesLoaderSelectionPolicyFilterAware(),
      reaperSelectionPolicy: EXUpdatesReaperSelectionPolicyFilterAware()
    )
  }
}
